cpp/src/arrow/filesystem/localfs.cc

```diff
@@ -309,6 +314,240 @@ Result<std::vector<FileInfo>> LocalFileSystem::GetFileInfo(const FileSelector& s
+        if (!init.ok()) {
+          return Finish(init);
+        }
+      }
+      while (idx_ < child_fns_.size()) {
+        auto full_fn = dir_fn_.Join(child_fns_[idx_++]);
+        auto res = StatFile(full_fn.ToNative());
+        if (!res.ok()) {
+          return Finish(res.status());
+        }
+
+        auto info = res.MoveValueUnsafe();
+
+        // Try to recurse into subdirectories, if needed.
+        if (info.type() == FileType::Directory &&
+            nesting_depth_ < selector_.max_recursion && selector_.recursive) {
+          auto status = DoDiscovery(std::move(full_fn), nesting_depth_ + 1, selector_,
+                                    file_gen_producer_, file_info_batch_size_);
+          if (!status.ok()) {
+            return Finish(status);
+          }
+        }
+        // Everything is ok. Add the item to the current chunk of data.
+        current_chunk_.emplace_back(std::move(info));
+        // Keep `current_chunk_` as large, as `batch_size_`.
+        // Otherwise, yield the complete chunk to the caller.
+        if (current_chunk_.size() == file_info_batch_size_) {
+          FileInfoVector yield_vec;
+          std::swap(yield_vec, current_chunk_);
+          const size_t items_left = child_fns_.size() - idx_;
+          current_chunk_.reserve(
+              items_left >= file_info_batch_size_ ? file_info_batch_size_ : items_left);
+          return yield_vec;
+        }
+      }  // while (idx_ < child_fns_.size())
+
+      // Flush out remaining items
+      if (!current_chunk_.empty()) {
+        return std::move(current_chunk_);
+      }
+      return Finish();
+    }
+
+   private:
+    /// Close the producer end of stream and return iteration end marker.
+    Result<FileInfoVector> Finish(Status status = Status::OK()) {
+      file_gen_producer_.Finish();
+      ARROW_RETURN_NOT_OK(status);
+      return IterationEnd<FileInfoVector>();
+    }
+  };
+
+  /// Create an instance of  `DiscoveryImplIterator` under the hood for the
+  /// specified directory, wrap it in the `BackgroundGenerator + TransferredGenerator`
+  /// bundle and feed the results to the main producer queue.
+  static Status DoDiscovery(const PlatformFilename& dir_fn, int32_t nesting_depth,
+                            FileSelector selector, AutoClosingProducer file_gen_producer,
+                            int32_t file_info_batch_size) {
+    ARROW_RETURN_IF(file_gen_producer.IsFinished(),
+                    arrow::Status::Cancelled("Discovery cancelled"));
+
+    ARROW_ASSIGN_OR_RAISE(
+        auto gen,
+        MakeBackgroundGenerator(Iterator<FileInfoVector>(DiscoveryImplIterator(
+                                    std::move(dir_fn), nesting_depth, std::move(selector),
+                                    file_gen_producer, file_info_batch_size)),
+                                io::default_io_context().executor()));
+    gen = MakeTransferredGenerator(std::move(gen), arrow::internal::GetCpuThreadPool());
```

Comment:
Hmm... why are we transferring to the CPU thread pool? That doesn't seem necessary.
cc @westonpace

Reply:
But, as far as I understand, if the generator is not transferred back, every continuation attached to every future produced by this generator, will also be run on IO thread pool? 

This is undesirable, because we would want to do arbitrary (possibly compute-intensive) stuff on each delivered chunk from `GetFileInfoGenerator()`.

Reply:
But you may also want to do IO on the `GetFileInfoGenerator` results, in which case you'll transfer back and forth between executors?

Reply:
Hmmm, looks reasonable. What we get from `GetFileInfoGenerator()` is only a raw file handle, so we would want to process it somehow later, e.g. read the file, inspect the metadata etc. Given that, the generator output is going to produce some more IO operations either way, in most cases.

I think `MakeTransferredGenerator()` is not necessary, indeed.

Reply:
I'll leave a note about that in `DoDiscovery()`, so that future readers don't have to guess.

Reply:
Strange thing: after removing a call to `MakeTransferredGenerator()`, the benchmark hangs. Debugging that...

In the meanwhile, maybe you know of any quirks when dealing with background generators, related to this problem?

Reply:
@westonpace would be the best person to answer that question.

Reply:
Okay, it seems, that I need to transfer from the `BackgroundGenerator` anyway. There is a bunch of related issues about deadlocking background generators, e.g. https://issues.apache.org/jira/browse/ARROW-13109 and https://issues.apache.org/jira/browse/ARROW-13110, which added a corresponding note to `MakeBackgroundGenerator`:

> You MUST transfer away from this background generator. Otherwise there could be a race condition if a callback on the background thread deletes the last consumer reference to the background generator. You can transfer onto the same executor as the background thread, it is only neccesary to create a new thread task, not to switch executors.

Though, it's legal to specify the same executor for a background generator, in order to start a new task but without switching to the other executor.

So, there should be `MakeTransferredGenerator`, but we can specify the same `io_executor` as a target.

Reply:
Sorry I wasn't much help, I've been trying to wrap up my work on a better async task scheduler.  I will give this PR a thorough look tomorrow.  What you are describing sounds correct.  Creating a new thread task doesn't seem ideal but it should work ok.  I agree that, in the ideal case, no transfer should be necessary and we should just remain on the I/O thread.